src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1080,34 +1077,94 @@ private bool TryReadNextLine(out ReadOnlySpan<byte> line)
+                    // lines, where the continuation line begins with a space or horizontal tab.
+                    // The feature was deprecated in RFC 7230 3.2.4, but some servers still use it.
+                    if (length > 0)
                     {
-                        ThrowInvalidHttpResponse();
+                        // If the newline is the last character we've buffered, we need at least
+                        // one more character in order to see whether it's space/tab, in which
+                        // case it's a folded header.
+                        if (lfIndex + 1 == _readLength)
+                        {
+                            // The LF is at the end of the buffer, so we need to read more
+                            // to determine whether there's a continuation.  We'll read
+                            // and then loop back around again, but to avoid needing to
+                            // rescan the whole header, reposition to one character before
+                            // the newline so that we'll find it quickly.
+                            int backPos = _readBuffer[lfIndex - 1] == '\r' ? lfIndex - 2 : lfIndex - 1;
+                            Debug.Assert(backPos >= 0);
+                            previouslyScannedBytes = backPos - _readOffset;
+                            _allowedReadLineBytes -= backPos - scanOffset;
+                            ThrowIfExceededAllowedReadLineBytes();
+                            await FillAsync().ConfigureAwait(false);
+                            continue;
+                        }
+
+                        // We have at least one more character we can look at.
+                        Debug.Assert(lfIndex + 1 < _readLength);
+                        char nextChar = (char)_readBuffer[lfIndex + 1];
+                        if (nextChar == ' ' || nextChar == '\t')
+                        {
+                            // The next header is a continuation.
+
+                            // Folded headers are only allowed within header field values, not within header field names,
+                            // so if we haven't seen a colon, this is invalid.
+                            if (Array.IndexOf(_readBuffer, (byte)':', _readOffset, lfIndex - _readOffset) == -1)
+                            {
+                                ThrowInvalidHttpResponse();
+                            }
+
+                            // When we return the line, we need the interim newlines filtered out. According
+                            // to RFC 7230 3.2.4, a valid approach to dealing with them is to "replace each
+                            // received obs-fold with one or more SP octets prior to interpreting the field
+                            // value or forwarding the message downstream", so that's what we do.
+                            _readBuffer[lfIndex] = (byte)' ';
```

Comment:
This shouldn't be necessary.  TryAddWithoutValidation can handle embedded newlines.  I.e. this works fine and produces the same result as if \r\n was not present:

```csharp
            var request = new HttpRequestMessage();
            request.Headers.TryAddWithoutValidation("Accept", "text/html,\r\n */*");
            foreach (var v in request.Headers.Accept)
            {
                Console.WriteLine(v);
            }
```

Reply:
Interesting.  Maybe it should?  Or if not TryAddWithoutValidation, perhaps when the value is retrieved?  There's a whole process here about when stuff is validated/parsed/normalized.  If we can just reuse that, it would be nice.  Let me look into this real quick.

Reply:
What's interesting is that we specifically check for this situation in HttpHeaders.ParseSingleRawHeaderValue.  A header value with "\r\n" followed by space or tab is considered valid; one with "\r\n" followed by anything else is not.

If you change the code above to 

```csharp
request.Headers.TryAddWithoutValidation("FooBar", "this is a \r\ntest");
```

Then you'll get an exception of GetValues("FooBar") saying that the FooBar header was not found.

Reply:
It seems to me that the validation logic in ParseSingleRawHeaderValue should remove the \r\n -- that seems most in keeping with what this logic is intended to do, which is to produce a validated header value.

Reply:
I'll open an issue for it.  Changing that shared logic is a much more impactful change than the localized one I have here, and I don't want to block this on that broader issue.